<a href="https://colab.research.google.com/github/Dami-Adey/DataScience_Project1/blob/main/uk_university_DataCleaning.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Data Import and Initialisation

In [68]:
# imports
import pandas as pd
import numpy as np
import seaborn as sns
from google.colab import files

In [69]:
# upload data
uploaded = files.upload()

Saving uk_universities.csv to uk_universities (2).csv


In [70]:
# initialise dataframe
df = pd.read_csv('uk_universities.csv')
df.head()

University_name              Region  Founded_year  \
0   University of Cambridge     East of England          1209   
1      University of Oxford  South East England          1096   
2  University of St Andrews            Scotland          1413   
3   Imperial College London              London          1907   
4   Loughborough University       East Midlands          1966   

                                               Motto  UK_rank  World_rank  \
0               From here, light and sacred draughts        1           4   
1                               The Lord is my light        2           2   
2                                      Ever to excel        3          86   
3  Knowledge is the adornment and safeguard of th...        4           8   
4                   With Truth, Knowledge and Labour        5         404   

   CWUR_score  Minimum_IELTS_score  UG_average_fees_(in_pounds)  \
0        94.1                  6.5                        21750   
1        93.3                  6.5                        21770   
2        75.8                  6.5                        17040   
3        86.6                  6.5                        23500   
4        72.8                  5.5                        16400   

   PG_average_fees_(in_pounds)  ... Student_satisfaction Student_enrollment  \
0                        23187  ...               85.50%      20,000-24,999   
1                        19888  ...               86.50%      25,000-29,999   
2                        15440  ...               87.90%      10,000-14,999   
3                        29900  ...               77.90%      15,000-19,999   
4                        16400  ...               85.80%      15,000-19,999   

  Academic_staff Control_type Academic_Calender Campus_setting  \
0     over-5,000       Public        Trimesters          Urban   
1     over-5,000       Public        Trimesters          Urban   
2    1,000-1,499       Public         Semesters       Suburban   
3    4,000-4,499       Public        Continuous          Urban   
4    1,500-1,999       Public         Semesters       Suburban   

  Estimated_cost_of_living_per_year_(in_pounds)  Latitude  Longitude  \
0                                         12000   52.2054     0.1132   
1                                         11500   51.7548    -1.2544   
2                                         12000   56.3417    -2.7943   
3                                         10700   51.4988    -0.1749   
4                                          9398   52.7650    -1.2321   

                Website  
0         www.cam.ac.uk  
1          www.ox.ac.uk  
2  www.st-andrews.ac.uk  
3          www.ic.ac.uk  
4      www.lboro.ac.uk/  

[5 rows x 21 columns]

In [71]:
# inspect data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 131 entries, 0 to 130
Data columns (total 21 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   University_name                                131 non-null    object 
 1   Region                                         131 non-null    object 
 2   Founded_year                                   131 non-null    int64  
 3   Motto                                          114 non-null    object 
 4   UK_rank                                        131 non-null    int64  
 5   World_rank                                     131 non-null    int64  
 6   CWUR_score                                     84 non-null     float64
 7   Minimum_IELTS_score                            131 non-null    float64
 8   UG_average_fees_(in_pounds)                    131 non-null    int64  
 9   PG_average_fees_(in_pounds)                    131 non

In [72]:
# dataframe shape
print(f'Dataset has {df.shape[0]} rows and {df.shape[1]} columns')

Dataset has 131 rows and 21 columns


# Data Cleaning

### Initial Comments and Insights

- Only one entry per university? 
  - If yes, are these all for the same year?
- Some null data in '**Motto**', '**CWUR_score**', '**Academic_Calendar**', and '**Campus_setting**'.
  - Review the significance of these features.
- Some numerical data currently represented as string objects.
  - Column indexs 10-16

## Only one entry per university?

In [73]:
df['University_name'].value_counts()

University of Cambridge           1
Abertay University                1
Glasgow Caledonian University     1
Birmingham City University        1
University of Gloucestershire     1
                                 ..
University of Reading             1
University of Sussex              1
University of Essex               1
University of Strathclyde         1
Ravensbourne University London    1
Name: University_name, Length: 131, dtype: int64

There are 131 unique university names. This matches the length of the dataframe suggesting that each university is represented only once each. 

## Missing Data

Now let's check to see how much data is missing from the set.

In [74]:
# check the number of null entries in each column
df.isnull().sum()

University_name                                   0
Region                                            0
Founded_year                                      0
Motto                                            17
UK_rank                                           0
World_rank                                        0
CWUR_score                                       47
Minimum_IELTS_score                               0
UG_average_fees_(in_pounds)                       0
PG_average_fees_(in_pounds)                       0
International_students                            0
Student_satisfaction                              0
Student_enrollment                                0
Academic_staff                                    0
Control_type                                      0
Academic_Calender                                26
Campus_setting                                   18
Estimated_cost_of_living_per_year_(in_pounds)     0
Latitude                                          0
Longitude   

Might be more intuitive to look at the percentage of data missing for each case.

In [75]:
# percentage view
df.isna().sum() * 100/ df.shape[0]

University_name                                   0.000000
Region                                            0.000000
Founded_year                                      0.000000
Motto                                            12.977099
UK_rank                                           0.000000
World_rank                                        0.000000
CWUR_score                                       35.877863
Minimum_IELTS_score                               0.000000
UG_average_fees_(in_pounds)                       0.000000
PG_average_fees_(in_pounds)                       0.000000
International_students                            0.000000
Student_satisfaction                              0.000000
Student_enrollment                                0.000000
Academic_staff                                    0.000000
Control_type                                      0.000000
Academic_Calender                                19.847328
Campus_setting                                   13.7404

4 columns with missing values. We'll start by dropping the 'Motto' column outright since it's quite an arbitrary feature. Ultimately we could circle back on this once our foundation in NLP is a bit stronger. For now, we'll assume that the motto has little analytical relevance.

### The '**Motto**' Column

In [76]:
# drop the motto column
df = df.drop('Motto', axis=1)
df.head()

University_name              Region  Founded_year  UK_rank  \
0   University of Cambridge     East of England          1209        1   
1      University of Oxford  South East England          1096        2   
2  University of St Andrews            Scotland          1413        3   
3   Imperial College London              London          1907        4   
4   Loughborough University       East Midlands          1966        5   

   World_rank  CWUR_score  Minimum_IELTS_score  UG_average_fees_(in_pounds)  \
0           4        94.1                  6.5                        21750   
1           2        93.3                  6.5                        21770   
2          86        75.8                  6.5                        17040   
3           8        86.6                  6.5                        23500   
4         404        72.8                  5.5                        16400   

   PG_average_fees_(in_pounds) International_students Student_satisfaction  \
0                        23187                 20.20%               85.50%   
1                        19888                 16.80%               86.50%   
2                        15440                 40.40%               87.90%   
3                        29900                 41.40%               77.90%   
4                        16400                 22.00%               85.80%   

  Student_enrollment Academic_staff Control_type Academic_Calender  \
0      20,000-24,999     over-5,000       Public        Trimesters   
1      25,000-29,999     over-5,000       Public        Trimesters   
2      10,000-14,999    1,000-1,499       Public         Semesters   
3      15,000-19,999    4,000-4,499       Public        Continuous   
4      15,000-19,999    1,500-1,999       Public         Semesters   

  Campus_setting  Estimated_cost_of_living_per_year_(in_pounds)  Latitude  \
0          Urban                                          12000   52.2054   
1          Urban                                          11500   51.7548   
2       Suburban                                          12000   56.3417   
3          Urban                                          10700   51.4988   
4       Suburban                                           9398   52.7650   

   Longitude               Website  
0     0.1132         www.cam.ac.uk  
1    -1.2544          www.ox.ac.uk  
2    -2.7943  www.st-andrews.ac.uk  
3    -0.1749          www.ic.ac.uk  
4    -1.2321      www.lboro.ac.uk/

### Miscellaneous Cleaning

Now we'll try and address the missing values in the remaining three columns. First, we'll get rid of any rows that have more than 2 or more missing features.

In [77]:
# confirm expected shape
df.shape

(131, 20)

In [78]:
# find dataframe rows with 2 or more missing features
df[df.isnull().sum(axis=1) >= 2].head()

University_name              Region  Founded_year  \
37              University of Sussex  South East England          1961   
41  University for the Creative Arts  South East England          2005   
42       Arts University Bournemouth  South West England          1885   
51         University of West London              London          1990   
62    Norwich University of the Arts     East of England          1845   

    UK_rank  World_rank  CWUR_score  Minimum_IELTS_score  \
37       38         140        77.7                  5.5   
41       42        2030         NaN                  4.5   
42       43        3420         NaN                  4.5   
51       52        2319         NaN                  4.5   
62       62        4300         NaN                  4.5   

    UG_average_fees_(in_pounds)  PG_average_fees_(in_pounds)  \
37                        14450                        14450   
41                        11870                        12150   
42                        12500                        12500   
51                        10650                        10995   
62                        12500                        13000   

   International_students Student_satisfaction Student_enrollment  \
37                 27.20%               76.40%      15,000-19,999   
41                 11.70%               78.00%        6,000-6,999   
42                 13.60%               82.90%        3,000-3,999   
51                 17.20%               84.70%      10,000-14,999   
62                  4.90%               77.70%        2,000-2,999   

   Academic_staff Control_type Academic_Calender Campus_setting  \
37    2,000-2,499       Public               NaN            NaN   
41        300-399       Public               NaN            NaN   
42        300-399       Public               NaN            NaN   
51    1,000-1,499       Public         Semesters            NaN   
62        200-299       Public               NaN          Urban   

    Estimated_cost_of_living_per_year_(in_pounds)  Latitude  Longitude  \
37                                          12500   50.8677    -0.0875   
41                                           8990   51.2151    -0.8054   
42                                           6000   50.7417    -1.8977   
51                                          10000   51.5069    -0.3032   
62                                           9415   52.6302     1.2967   

                Website  
37     www.sussex.ac.uk  
41  www.ucreative.co.uk  
42            aub.ac.uk  
51        www.uwl.ac.uk  
62       www.nua.ac.uk/

In [79]:
# find indices
twoEmpty_indx = df.index[df.isnull().sum(axis=1) >= 2]
print(twoEmpty_indx)

Int64Index([ 37,  41,  42,  51,  62,  70,  71,  75,  79,  80,  84,  86,  91,
            102, 103, 109, 114, 115, 120, 128, 129],
           dtype='int64')


In [80]:
# drop rows
df = df.drop(twoEmpty_indx, axis=0)

In [81]:
# confirm expected shape
df.shape

(110, 20)

Now let's have another look at the missing data:

In [82]:
# check the number of null entries in each column
df.isnull().sum()

University_name                                   0
Region                                            0
Founded_year                                      0
UK_rank                                           0
World_rank                                        0
CWUR_score                                       28
Minimum_IELTS_score                               0
UG_average_fees_(in_pounds)                       0
PG_average_fees_(in_pounds)                       0
International_students                            0
Student_satisfaction                              0
Student_enrollment                                0
Academic_staff                                    0
Control_type                                      0
Academic_Calender                                 6
Campus_setting                                    8
Estimated_cost_of_living_per_year_(in_pounds)     0
Latitude                                          0
Longitude                                         0
Website     

We're still missing quite a few records from '**CWUR_score**', so we'll drop those as well.

In [83]:
# drop any outstanding rows with no CWUR_score.
df = df.drop(df.index[df['CWUR_score'].isna()], axis=0)

We'll also drop the outstanding '**Academic_Calender**' and '**Campus_setting**' data. At this point, we **could** do some data sourcing to see if we can fill in the missing data manually, but for now, we'll proceed by just removing the records with data missing. If model performance isn't sufficient, we can re-assess the data removal at this stage.

In [84]:
# remove records with missing academic calendar status
df = df.drop(df.index[df['Academic_Calender'].isna()], axis=0)
# remove entries with missing campus setting status
df = df.drop(df.index[df['Campus_setting'].isna()], axis=0)

In [85]:
# confirm operation
df.isnull().sum()

University_name                                  0
Region                                           0
Founded_year                                     0
UK_rank                                          0
World_rank                                       0
CWUR_score                                       0
Minimum_IELTS_score                              0
UG_average_fees_(in_pounds)                      0
PG_average_fees_(in_pounds)                      0
International_students                           0
Student_satisfaction                             0
Student_enrollment                               0
Academic_staff                                   0
Control_type                                     0
Academic_Calender                                0
Campus_setting                                   0
Estimated_cost_of_living_per_year_(in_pounds)    0
Latitude                                         0
Longitude                                        0
Website                        

And with that all the missing data has been addressed. There are still some issues we'd like to address. For one, there are still several numerical data fields currently instantiated as string objects. We'll treat those next. 

# Feature Cleaning

Reminding ourselves of the dataframe construction at this point. We should now have 68 entries remaining.
- Some concerns that we might have cut too much data, but we'll push on for now. 

In [86]:
# datatype refresher
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 68 entries, 0 to 122
Data columns (total 20 columns):
 #   Column                                         Non-Null Count  Dtype  
---  ------                                         --------------  -----  
 0   University_name                                68 non-null     object 
 1   Region                                         68 non-null     object 
 2   Founded_year                                   68 non-null     int64  
 3   UK_rank                                        68 non-null     int64  
 4   World_rank                                     68 non-null     int64  
 5   CWUR_score                                     68 non-null     float64
 6   Minimum_IELTS_score                            68 non-null     float64
 7   UG_average_fees_(in_pounds)                    68 non-null     int64  
 8   PG_average_fees_(in_pounds)                    68 non-null     int64  
 9   International_students                         68 non-n

The problematic features occupy the column indices 9-12. After inspecting the data, it's clear that these columns require numerical datatypes.

## '**International_students**' and '**Student_satisfaction**' Columns

Both of these columns contain strings ending with a percentage symbol. We'll have to define a function to reformat the string data so that we can get rid of the percentage symbol before converting to a float.

In [94]:
# function to remove percentage sign from a string
def remove_percent(str):
  return ''.join([char for char in str if '%' not in char])

In [121]:
# confirm string structure
df['International_students']

0      20.20%
1      16.80%
2      40.40%
3      41.40%
4      22.00%
        ...  
107     5.20%
110    11.10%
118    21.10%
119    10.40%
122     5.70%
Name: International_students, Length: 68, dtype: object

In [122]:
# pull out the percentage sign and convert to the appropriate numeric datatype
df['International_students'] = pd.to_numeric(df['International_students'].apply(lambda x: remove_percent(x)))
df['Student_satisfaction'] = pd.to_numeric(df['Student_satisfaction'].apply(lambda x: remove_percent(x)))
# confirm expected operation
print(df[['International_students','Student_satisfaction']])

     International_students  Student_satisfaction
0                      20.2                  85.5
1                      16.8                  86.5
2                      40.4                  87.9
3                      41.4                  77.9
4                      22.0                  85.8
..                      ...                   ...
107                     5.2                  78.0
110                    11.1                  76.8
118                    21.1                  75.1
119                    10.4                  79.0
122                     5.7                  79.3

[68 rows x 2 columns]


## '**Student_enrollment**' Column

For the '**Student_enrollment**' column, we'll need slightly different reformating protocol.

In [123]:
# check string format
df['Student_enrollment']

0      20,000-24,999
1      25,000-29,999
2      10,000-14,999
3      15,000-19,999
4      15,000-19,999
           ...      
107    15,000-19,999
110    20,000-24,999
118    15,000-19,999
119    20,000-24,999
122    20,000-24,999
Name: Student_enrollment, Length: 68, dtype: object

Our reformatting function here will separate the strings - which represent a range of values - into upper and lower bounds. After converting both numbers to floats, we can compute a simple average and return the processed data to the dataframe.
- Note: the function will also have to remove the commmas 

In [144]:
# calculates an avergae using the upper and lower bounds of a defined range
def average_from_range(str):
  # remove all commas
  commas_removed = str.replace(',','')
  # split on the hyphen, to get the upper and lower bounds
  lower_bound = float(commas_removed.split('-')[0])
  upper_bound = float(commas_removed.split('-')[1])
  # return the average
  return (lower_bound + upper_bound) / 2

Now we'll apply the transformation.
- Note: Conversion to a numeric datatype happens internally within the function space when evaluating the average, so we don't need an explcit 'to_numeric' call on the dataframe.

In [145]:
# convert ranges to averages
df['Student_enrollment'] = df['Student_enrollment'].apply(lambda x: average_from_range(x))
# confirm expected operation
print(df['Student_enrollment'])

0      22499.5
1      27499.5
2      12499.5
3      17499.5
4      17499.5
        ...   
107    17499.5
110    22499.5
118    17499.5
119    22499.5
122    22499.5
Name: Student_enrollment, Length: 68, dtype: float64


## '**Academic_staff**' Column

In [127]:

test = '20,000-24,999'

In [142]:
a, b = test.replace(',','').split('-')

TypeError: ignored

In [135]:
a

'20000'

In [141]:
float('20.23')

20.23